### PR TITLE
modules/hardware/all-firmware: add enableAllUnfreeFirmware

### PR DESCRIPTION
**Pull request by Profpatsch**
cc @Mic92

### CODE REVIEW

File: nixos/modules/hardware/all-firmware.nix [hunk 1]
```diff
@@ -12,7 +12,14 @@ with lib;
       default = false;
       type = types.bool;
       description = ''
-        Turn on this option if you want to enable all the firmware shipped in linux-firmware.
+        Turn on this option if you want to enable all the firmware shipped in linux-firmware. For unfree firmware, see <literal>hardware.enableAllUnfreeFirmware</literal>.
+      '';
+    };
+    hardware.enableAllUnfreeFirmware = mkOption {
```
Comment (Mic92) [inline nixos/modules/hardware/all-firmware.nix]: Would it not been possible to just check if allowUnfree was true?

**Comment by fpletz** [comment]
Reopen because there is a dispute about the fix in https://github.com/NixOS/nixpkgs/commit/cc75ccbf64bf8c874cd884cb5c1534526553ee6c.

**Comment by Profpatsch** [comment]
@Mic92 I’d prefer for it to be explicit. By automatically including it when `allowUnfree` is set you suddenly depend on impurities. It will be very hard to debug why your system stopped working when you accidentally rebuild it with `allowUnfree` disabled once.

The more I think about it, the scarier it gets. I imagine removing some unfree package like `Skype`, thinking that I can remove the need for unfree packages, suddenly my wifi stops working and I *have no idea why*. Near un-debuggable.

**Comment by Mic92** [comment]
But then the other option naming is also misleading `enableAllFirmware` does not enable all firmware.

**Comment by Profpatsch** [comment]
> enableAllFirmware does not enable all firmware

That’s why I added a reference to the documentation. We should always handle unfree software explicitely and separately imho.

File: nixos/modules/hardware/all-firmware.nix [hunk 2]
```diff
@@ -12,7 +12,14 @@ with lib;
       default = false;
       type = types.bool;
       description = ''
-        Turn on this option if you want to enable all the firmware shipped in linux-firmware.
+        Turn on this option if you want to enable all the firmware shipped in linux-firmware. For unfree firmware, see <literal>hardware.enableAllUnfreeFirmware</literal>.
```
Comment (Mic92) [inline nixos/modules/hardware/all-firmware.nix]: For non-redistributal firmware, firmwareLinuxNonfree is also non-free.

**Comment by Mic92** [comment]
All firmware files in this list are proprietary.

**Comment by Mic92** [comment]
I made this pull request to solve both concerns: https://github.com/NixOS/nixpkgs/pull/25636

**Pull request closed without merging**